idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/references/FirReferenceResolveHelper.kt

```diff
@@ -374,42 +375,104 @@ internal object FirReferenceResolveHelper {
+            while (true) {
+                if (expression === qualifiedAccess.selectorExpression) {
+                    return listOfNotNull(classId.toTargetPsi(session, symbolBuilder))
+                }
+                val outerClassId = classId.outerClassId
+                val receiverExpression = qualifiedAccess.receiverExpression
+                if (receiverExpression !is KtDotQualifiedExpression) {
+                    // If the receiver is not a KtDotQualifiedExpression, it means we are hitting the end of nested receivers.
+                    if (receiverExpression == expression) {
+                        // If there is still an outer class, then return symbol of that class
+                        outerClassId?.let { return listOfNotNull(it.toTargetPsi(session, symbolBuilder)) }
+                        // Otherwise, it should be a package, so we return that
+                        return listOfNotNull(symbolBuilder.createPackageSymbolIfOneExists(classId.packageFqName))
+                    } else {
+                        // This is unexpected. The code probably contains some weird structures. In this case, we just fail the resolution
+                        // with zero results.
+                        return emptyList()
+                    }
+                }
+                qualifiedAccess = receiverExpression
+                if (outerClassId == null) break
+                classId = outerClassId
             }
-            val receiverClassId = if (parent.receiverExpression == expression) {
-                /*
-                 * <caret>A.Named.i -> class A
-                 */
-                val name = fir.relativeClassFqName?.pathSegments()?.firstOrNull()
-                name?.let { ClassId(fir.packageFqName, it) }
-            } else null
-            val parentFir = selectorExpression.getOrBuildFir(analysisSession.firResolveState)
-            when {
-                parentFir is FirQualifiedAccess -> {
-                    return listOfNotNull(
-                        (receiverClassId ?: classId).toTargetPsi(session, symbolBuilder, parentFir.calleeReference)
-                    )
+
+            // Handle package names
+            var packageFqName = classId.packageFqName
+
+            while (!packageFqName.isRoot) {
+                if (expression === qualifiedAccess.selectorExpression) {
+                    return listOfNotNull(symbolBuilder.createPackageSymbolIfOneExists(packageFqName))
                 }
-                receiverClassId != null -> {
-                    return listOfNotNull(receiverClassId.toTargetPsi(session, symbolBuilder))
+                val parentPackageFqName = packageFqName.parent()
+                val receiverExpression = qualifiedAccess.receiverExpression
+                if (receiverExpression !is KtDotQualifiedExpression) {
+                    // If the receiver is not a KtDotQualifiedExpression, it means we are hitting the end of nested receivers.
```

Comment:
Copy-n-pasted? You meant "we are hitting the top package"?

Reply:
No, this comment as it stands is what I meant. By "nested receivers" I am describing the PSI structure. But I can add more comments about this end must be referring to a root package.